ocaml/typing/typecore.ml

```diff
@@ -504,17 +505,17 @@ let mode_exact mode =
   { (mode_default mode) with
     exact = true }
 
-let mode_argument ~funct ~index ~position ~partial_app alloc_mode =
+let mode_argument ~funct ~index ~position_and_mode ~partial_app alloc_mode =
   let vmode = Value.of_alloc alloc_mode in
   if partial_app then mode_default vmode
-  else match funct.exp_desc, index, (position : apply_position) with
+  else match funct.exp_desc, index, position_and_mode.apply_position with
   | Texp_ident (_, _, {val_kind =
       Val_prim {Primitive.prim_name = ("%sequor"|"%sequand")}},
                 Id_prim _, _), 1, Tail ->
-     (* The second argument to (&&) and (||) is in
-        tail position if the call is *)
-      (* vmode is wrong; fine because of mode crossing on boolean *)
-     mode_return vmode
+     (* The second argument to (&&) and (||) is a boolean and crosses modes, so
+        we expect [mode_max]. It is also at the same function tail position if
+        the call is *)
+     mode_max_with_position (RTail (Option.get position_and_mode.region_mode, FTail))
```

Comment:
I'd prefer we didn't make assumptions about the type of `&&` and `||` only their runtime behaviour. This matches what we do elsewhere for known primitives. Also, if we add more modes later that `bool` doesn't cross then this assumption will be wrong. So I'd rather we tried to give the "correct" mode and let mode crossing handle weakening it based on type.

Reply:
Besides, although it likely doesn't happen anywhere, it could be technically possible to declare `&&` with type `bool -> 'a option -> 'a option`; if might be worth thinking what would happen in this case.

Reply:
That makes sense. 

Just to confirm: there are other similar code.

- the lower and upper bound of `Pexp_loop` are int and checked with `mode_max`. Apparantly your arguments apply to this as well and we should fix this. 

- the body of a loop is checked with `mode_max` - I guess this is fine because the value if any are discarded anyway.

Reply:
@Ekdohibs I think here it's specifically about `%sequor` and `%sequand`; I.e., the primitives. The meaning of those are quite fixed.

Reply:
@riaqn yes, I meant declaring `%sequand` with that type, sorry for the confusion! If I'm not mistaken, the runtime semantics of `%sequand` would be compatible with that type.
Concerning `Pexp_loop`, it is different, because it is a language construct with a special syntax, and not a primitive (declared with `external ( && ) : bool -> bool -> bool = "%sequand"`)

Reply:
The main difference is that primtives are declared with a type, which we trust in the rest of the compiler (this allows us to define `Obj.magic`!), while language constructs have a special syntax, are not user-declarable and so their types are always known in the rest of the compiler.

Reply:
I think there are two separate concerns, and mixing them is causing some confusion in this thread.

1. The relationship between types and `max_mode` may change in the future, which may indeed mean we should adjust the code related to `Pexp_loop`.
2.  We don't actually know what type the `%sequand` primitive is declared with, and as @Ekdohibs points out there are even some non-standard choices that are probably safe to use right now, which means that we should try not to rely on that assumption.

Reply:
Ah, thanks for clearing that out! I understand the part about `Pexp_loop` better now.

Reply:
The mode crossing bit is fixed now.

For the other thing,  Leo and me discussed and decides that they don't need fixing.